File: zbestcomic/src/main/java/vn/loitp/app/activity/comicinfo/chap/ChapAdapter.java
package vn.loitp.app.activity.comicinfo.chap;

/**
 * Created by www.muathu@gmail.com on 12/8/2017.
 */

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;

import java.util.List;

import loitp.basemaster.R;
import vn.loitp.app.model.chap.Chap;

public class ChapAdapter extends RecyclerView.Adapter<ChapAdapter.ChapHolder> {

    public interface Callback {
        public void onClick(Chap chap, int position);

        public void onLongClick(Chap chap, int position);

        public void onClickDownload(Chap chap, int position);

        public void onLoadMore();
    }

    private Callback callback;

    private List<Chap> chapList;

    public class ChapHolder extends RecyclerView.ViewHolder {
        public Button btChap;
        public ImageView btDownload;

        public ChapHolder(View view) {
            super(view);
            btChap = (Button) view.findViewById(R.id.bt_chap);
            btDownload = (ImageView) view.findViewById(R.id.bt_download);
        }
    }

    public ChapAdapter(List<Chap> chapList, Callback callback) {
        this.chapList = chapList;
        this.callback = callback;
    }

    @Override
    public ChapHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_comic_chap, parent, false);
        return new ChapHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ChapHolder holder, int position) {
        Chap chap = chapList.get(position);
        holder.btChap.setText(chap.getTit());
        holder.btChap.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (callback != null) {
                    callback.onClick(chap, position);
                }
            }
        });
        holder.btChap.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                if (callback != null) {
                    callback.onLongClick(chap, position);
                }
                return true;
            }
        });
        holder.btDownload.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (callback != null) {
                    callback.onClickDownload(chap, position);
                }
            }
        });
        if (position == chapList.size() - 1) {
            if (callback != null) {
                callback.onLoadMore();
            }
        }
    }

    @Override
    public int getItemCount() {
        return chapList.size();
    }
}

File: core/src/main/java/com/core/helper/mup/comic/adapter/ChapAdapter.kt
package com.core.helper.mup.comic.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.R
import com.annotation.LogTag
import com.core.adapter.AnimationAdapter
import com.core.helper.mup.comic.model.Chap
import com.views.setSafeOnClickListener
import kotlinx.android.synthetic.main.view_row_comic_chap.view.*

@LogTag("CategoryAdapter")
class ChapAdapter : AnimationAdapter() {

    private val listChap = ArrayList<Chap>()

    fun setListData(listChap: List<Chap>) {
        this.listChap.addAll(listChap)
        notifyDataSetChanged()
    }

    var onClickRoot: ((Chap) -> Unit)? = null

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(chap: Chap) {
            itemView.tvChap.text = chap.title
            itemView.tvChap.setSafeOnClickListener {
                onClickRoot?.invoke(chap)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            ViewHolder(LayoutInflater.from(parent.context).inflate(
                    R.layout.view_row_comic_chap, parent,
                    false
            ))

    override fun getItemCount(): Int = listChap.size

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is ViewHolder) {
            holder.bind(chap = listChap[position])
        }
    }

}
